src/core/stream/initial_seek_and_play.ts

```diff
@@ -84,38 +110,52 @@ function doInitialSeek(
  * @returns {object}
  */
 export default function seekAndLoadOnMediaEvents(
+  clock$ : Observable<IStreamClockTick>,
   mediaElement : HTMLMediaElement,
   startTime : number|(() => number),
   mustAutoPlay : boolean
-) : {
-  seek$ : Observable<void>;
-  load$ : Observable<"autoplay-blocked"|"autoplay"|"loaded">;
-} {
-  const seek$ = doInitialSeek(mediaElement, startTime);
-
-  const load$ = canPlay(mediaElement).pipe(
-
-    tap(() => log.info("Stream: canplay event")),
+  ) : { seek$ : Observable<void>; load$ : Observable<ILoadEvents> } {
+  const seek$ = onLoadedMetadata$(mediaElement).pipe(
+    mergeMap(() => doInitialSeek(mediaElement, startTime)),
+    // equivalent to a sane shareReplay:
+    // https://github.com/ReactiveX/rxjs/issues/3336
+    // XXX TODO Replace it when that issue is resolved
+    multicast(() => new ReplaySubject(1)),
+    refCount()
+  );
 
+  const load$ = seek$.pipe(
     mergeMap(() => {
-      if (mustAutoPlay) {
-        return play$(mediaElement).pipe(
-          mapTo("autoplay" as "autoplay"),
-          catchError((error) => {
-            if (error.name === "NotAllowedError") {
-              // auto-play was probably prevented.
-              log.warn("Stream: Media element can't play." +
-                " It may be due to browser auto-play policies.");
-              return observableOf("autoplay-blocked" as "autoplay-blocked");
-            } else {
-              throw error;
-            }
-          })
-        );
-      }
-      return observableOf("loaded" as "loaded");
-    }),
+      const event = (() => {
+        if (mustAutoPlay) {
+          return play$(mediaElement).pipe(
+            mapTo("autoplay" as "autoplay"),
+            catchError((error) => {
+              if (error.name === "NotAllowedError") {
+                // auto-play was probably prevented.
+                log.warn("Stream: Media element can't play." +
+                  " It may be due to browser auto-play policies.");
+                return observableOf("autoplay-blocked" as "autoplay-blocked");
+              } else {
+                throw error;
+              }
+            })
+          );
+        }
+        return observableOf("loaded" as "loaded");
+      })();
 
+      return observableConcat(
+        canPlay(clock$, mediaElement).pipe(
+          tap(() => log.info("Stream: Can begin to play content")),
+          ignoreElements()
+        ),
+        observableOf(mediaElement.duration ?
+          "loaded-metadata" as "loaded-metadata" :
+          "not-loaded-metadata" as "not-loaded-metadata"),
+        event
```

Comment:
So we still try to auto-play even if we do not have the duration and thus we are not loaded?

Reply:
We consider not to have loaded metadata. However, the browser has emitted `hasLoadedMetadata` and `canPlay` events... So I assume the browser has told us that we can trust him to play.

In the typical case of Samsung Internet, even if the media duration is still set to 0, the browser has emitted events and the stream plays without problems. We should still rely on events to start playing.

The player states will still evolve in the same way.